### PR TITLE
Add mortar classes

### PR DESCRIPTION
**Pull request by peterrum**
### Description

Add classes useful for doing mortar FEM on sliding interfaces.

### Testing

A couple of unit tests. Some of them solve Poisson and Stokes equations and use multigrid.

### Documentation

--

### Miscellaneous (will be removed when merged)

--

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: tests/mortar/projection_matrix_01.cc [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
+
+using namespace dealii;
+
+template <int dim>
+Quadrature<dim>
+scale(const Quadrature<dim> &quadrature_2, const double alpha)
```
Comment (brunaccampos) [inline tests/mortar/projection_matrix_01.cc:18]: not sure I understand the meaning of alpha?

File: include/core/mortar_coupling_manager.h [hunk 2]
```diff
@@ -0,0 +1,1244 @@
...
+  std::vector<std::tuple<std::vector<double>,
+                         typename Triangulation<dim>::active_cell_iterator,
+                         std::vector<Point<dim>>,
+                         typename Triangulation<dim>::active_cell_iterator,
+                         std::vector<Point<dim>>,
+                         std::vector<Tensor<1, dim, Number>>,
+                         Number>>
+    all_intersections;
+
+  Number panalty_factor;
```
Comment (brunaccampos) [inline include/core/mortar_coupling_manager.h:504]:
```suggestion
  Number penalty_factor;
```

File: include/core/mortar_coupling_manager.h [hunk 3]
```diff
@@ -0,0 +1,1244 @@
...
+  void
+  add_sparsity_pattern_entries(TrilinosWrappers::SparsityPattern &dsp) const;
+
+  void
+  add_system_matrix_entries(
+    TrilinosWrappers::SparseMatrix &system_matrix) const;
+
+private:
+  Number
+  compute_pentaly_factor(const unsigned int degree, const Number factor) const;
```
Comment (brunaccampos) [inline include/core/mortar_coupling_manager.h:480]:
```suggestion
  compute_penalty_factor(const unsigned int degree, const Number factor) const;
```

File: include/core/mortar_coupling_manager.h [hunk 4]
```diff
@@ -0,0 +1,1244 @@
...
+  Number volume = 0;
+  for (unsigned int q = 0; q < quadrature.size(); ++q)
+    volume += fe_values.JxW(q);
+
+  Number surface_area = 0;
+  for (const auto f : cell->face_indices())
+    {
+      fe_face_values.reinit(cell, f);
+      const Number factor =
+        (cell->at_boundary(f) && !cell->has_periodic_neighbor(f)) ? 1. : 0.5;
```
Comment (brunaccampos) [inline include/core/mortar_coupling_manager.h:783]: Why do you have to check has_periodic_neighbor?

**Review by brunaccampos — CHANGES_REQUESTED** [review]
Great work! Just a couple typos and some questions

File: include/core/mortar_coupling_manager.h [hunk 5]
```diff
@@ -0,0 +1,1244 @@
...
+#include <deal.II/matrix_free/fe_evaluation.h>
+#include <deal.II/matrix_free/fe_point_evaluation.h>
+#include <deal.II/matrix_free/matrix_free.h>
+#include <deal.II/matrix_free/tools.h>
+
+using namespace dealii;
+
+template <int dim>
+double
+point_to_rad(const Point<dim> &point)
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/core/mortar_coupling_manager.h:26]: At the moment, this is the case. One could generalize this in the future once we need other criteria.

File: tests/mortar/manager_02.cc [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+{
+  const unsigned int dim                 = 2;
+  const unsigned int n_subdivisions      = 10;
+  const unsigned int n_quadrature_points = 3;
+  const double       radius              = 1.2;
+
+  const double delta = 2 * numbers::PI / n_subdivisions;
+
+  for (const double scale :
+       {0.0, 0.1, 0.5, 0.9, 1.1, 2.1, n_subdivisions - 0.9})
```
[the remark this answers is not in the document]
Comment (peterrum) [reply tests/mortar/manager_02.cc:21]: > Are these just random values?

The values are random; but try to test different configurations. Rotated by 0, 1, 2 cells ...

> Is there a reason for the jump from 2.1 to n_subdivisions - 0.9 = 9.1?

This will trigger a wrap around of the interval: `[x, 2pi] \caup [0, y]`.

File: tests/mortar/data_exchange_01.cc [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+          {
+            const auto indices = mm.get_indices(point_to_rad(face->center()));
+            const auto points  = mm.get_points(point_to_rad(face->center()));
+
+            for (unsigned int ii = 0; ii < indices.size(); ++ii)
+              {
+                unsigned int i = indices[ii];
+                unsigned int id_local, id_ghost;
+
+                if (face->boundary_id() == 0)
```
[the remark this answers is not in the document]
Comment (peterrum) [reply tests/mortar/data_exchange_01.cc:97]: This is only the case in this test. In the main code, which contains more or less the same lines as in this test, once can specify the boundary id.

File: tests/mortar/projection_matrix_01.cc (hunk 1 above)
Comment (peterrum) [reply tests/mortar/projection_matrix_01.cc:18]: It is just a scaling factor. I can rename it.

File: include/core/mortar_coupling_manager.h (hunk 4 above)
Comment (peterrum) [reply include/core/mortar_coupling_manager.h:783]: To be general. In deal.II, periodic faces are at the boundary and are not treated as internal faces, which would be the right thing for the computation of the penalty parameter.

File: tests/mortar/manager_02.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (brunaccampos) [reply tests/mortar/manager_02.cc:21]: okay, makes sense!

File: tests/mortar/projection_matrix_01.cc [hunk 8]
```diff
@@ -15,11 +15,11 @@ using namespace dealii;
...
 Quadrature<dim>
-scale(const Quadrature<dim> &quadrature_2, const double alpha)
+scale(const Quadrature<dim> &quadrature_2, const double factor)
 {
   std::vector<Point<dim>> points;
   for (const auto p : quadrature_2.get_points())
-    points.emplace_back(p * alpha);
+    points.emplace_back(p * factor);
   return Quadrature<dim>(points);
 }
```
Comment (brunaccampos) [inline tests/mortar/projection_matrix_01.cc:25]: I think now some places in the test have 'alpha' and others are named 'factor'. I just wasn't sure if it had a specific meaning, but no need to rename!

**Review by brunaccampos — APPROVED** [review]

**Comment by brunaccampos** [comment]
I added some descriptions throughout the code just to help me understand what was being done. Please feel free to change anything that doesn't make sense!

**Comment by blaisb** [comment]
@brunaccampos can you rebase? Then I would merge it afterwards

**Review by blaisb — APPROVED** [review]

**Comment by peterrum** [comment]
Failing test is unrelated:

```
The following tests FAILED:
	 68 - lethe-particles/insert_file_3d.mpirun=2.debug (Failed)
```
